### PR TITLE
DORA Lead Time to Change

### PR DESCRIPTION
**Pull request by jmelis**
As part of the Services Dashboard effort, the DORA LTTC (lead time to change metrics) must be collected.

The purpose of this MR is to enhance DashdotDB with the capability of ingesting information about the LTTC of the commits associated with deployment events.

Ticket: [OSDEV-1255](https://issues.redhat.com/browse/OSDEV-1255)

The ERD has been updated to reflect the additional database tables, the swagger file changes reflect the new additional methods.

This MR will be complemented by a corresponding MR to qontract-reconcile to collect information from the deployments and calling these new API endpoints to store them in DashdotDB.

Finally, this information will be rendered in Grafana, using the SQL datasource of DashdotDB.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ make test-data
 Run a PostgreSQL instance:
 
 ```
-docker run --rm -it -p 5432:5432 -e POSTGRES_PASSWORD=postgres postgres
+docker run -d --rm --name dashdot-postgres -p 5432:5432 -e POSTGRES_PASSWORD=postgres postgres
```
Comment (apahim) [inline README.md:34]: then we miss the console output, please add `docker log dashdot-postgres` as an additional command to show the console outputs :)

Comment (jmelis) [reply README.md:34]: ✅

Comment (jmelis) [reply README.md:34]: done

File: dashdotdb/services/dora.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    finish_timestamp=finish_timestamp,
+                    trigger_reason=dep_data["trigger_reason"],
+                    app_name=dep_data["app_name"],
+                    env_name=dep_data["env_name"],
+                    pipeline=dep_data["pipeline"],
+                )
+                db.session.add(deployment)
+                db.session.commit()
+
+                for commit_data in dep_data["commits"]:
```
Comment (apahim) [inline dashdotdb/services/dora.py:72]: commits is not required in the schema, meaning it can be None here and break the for loop.

**Review by apahim — APPROVED** [review]
lgtm. A couple of nitpicks, not blockers.

File: dashdotdb/services/dora.py (hunk 2 above)
Comment (jmelis) [reply dashdotdb/services/dora.py:72]: fixed, added it as required in swagger. Can you confirm this is good? Should I also do `dep_data.get('commits', [])` or no need?

File: migrations/versions/6d6661369d50_.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    #     column_name='data_type',
+    #     nullable=True,
+    #     type_=enum_type,
+    #     postgresql_using='data_type::datatypes'
+    # )
+    # op.execute('DROP TYPE datatypes_old')
+    #
+    # Gives this error:
+    # ERROR:  cannot cast type datatypes_old to datatypes
+
```
Comment (reavessm) [inline migrations/versions/6d6661369d50_.py]: This can probably be deleted

**Review by reavessm — APPROVED** [review]

File: dashdotdb/services/dora.py (hunk 2 above)
Comment (apahim) [reply dashdotdb/services/dora.py:72]: nope, this is fine as is

File: dashdotdb/controllers/dora.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def post(user, body) -> Union[tuple[str, int], tuple[DORAInsertStats, int]]:
+    try:
+        stats = DORA().insert(token=user, manifest=body)
+    except TokenNotFound:
+        return TOKEN_NOT_FOUND_MSG, TOKEN_NOT_FOUND_CODE
+
+    # if there's at least one entry created (or marked as duplicate), then this
+    # the post has been successful, but we report the errors.
```
Comment (geoberle) [inline dashdotdb/controllers/dora.py:18]: grammar nit :)

**Review by geoberle — APPROVED** [review]
LGTM

**Pull request merged**